File: day10/aoc2025_10.ts
import { readFileSync } from "node:fs";
import { init } from "z3-solver";

class Machine {
    lights: boolean[]
    buttons: number[][]
    joltages: number[]
    constructor(line: string) {
        const li = line.split(/\s+/)
        this.lights = [...li[0].slice(1, -1)].map(x => x === '#')
        this.buttons = li.slice(1, -1)
            .map(s => s.slice(1, -1).split(',').map(Number))
        this.joltages = li[li.length - 1].slice(1, -1).split(',').map(Number)
    }
}
function* combinations<T>(arr: T[], k: number, start = 0): Generator<T[]> {
    if (k === 0) {
        yield [];
        return;
    }
    for (let i = start; i <= arr.length - k; i++) {
        for (const rest of combinations(arr, k - 1, i + 1)) {
            yield [arr[i], ...rest];
        }
    }
}
function solve_subsets(machine: Machine): number {
    for (let subset_size = 1; subset_size < machine.buttons.length + 1; subset_size++) {
        for (let buttons_to_push of combinations(machine.buttons, subset_size)) {
            const lights = Array(machine.lights.length).fill(false)
            for (const button of buttons_to_push) {
                for (const b of button) lights[b] = !lights[b]
            }
            if (lights.every((v, i) => v === machine.lights[i]))
                return subset_size
        }
    }
}
const machines = readFileSync("day10/input10", "utf-8")
    .trim()
    .split("\n")
    .map(s => new Machine(s));

console.log('Part 1:', machines.map(solve_subsets).reduce((x, y) => x + y))

const z3Init = init();
async function solve_z3(machine: Machine): Promise<number> {
    const { Context } = await z3Init;
    const z3 = Context('main');
    const solver = new z3.Optimize();
    const numButtons = machine.buttons.length;
    const x = Array.from({ length: numButtons }, (_, i) => z3.Int.const(`x${i}`));
    for (const xi of x) solver.add(xi.ge(0));
    for (let row = 0; row < machine.joltages.length; row++) {
        const terms = x.filter((_, col) => machine.buttons[col].includes(row));
        const sum = terms.length === 1 ? terms[0] : z3.Sum(terms[0], ...terms.slice(1));
        solver.add(sum.eq(machine.joltages[row]));
    }
    solver.minimize(z3.Sum(x[0], ...x.slice(1)));
    await solver.check();
    const model = solver.model();
    return x.reduce((acc, xi) => acc + Number(model.eval(xi).toString()), 0);
}
Promise.all(machines.map(solve_z3)).then(results => {
    console.log('Part 2:', results.reduce((x, y) => x + y));
});

File: day10/aoc2025_10.scala
import com.microsoft.z3.*

class Machine(line: String) {
    private val li = line.split("\\s+")
    val lights = li.head.slice(1, li.head.length()-1).map(_ == '#').toArray
    val buttons = li.slice(1, li.length -1)
        .map(s => s.slice(1, s.length -1).split(',').map(_.toInt))
    val joltages = li.last.slice(1, li.last.length-1).split(',').map(_.toInt)
    override def toString(): String =
        def fmt[T](arr: Array[T]): String = arr.mkString("[", ",", "]")
        s"""
        |lights: ${fmt(lights)}
        |buttons: ${fmt(buttons.map(fmt))}
        |joltages: ${fmt(joltages)}""".stripMargin
}
def solve_subsets(lights_target: Array[Boolean], buttons: Array[Array[Int]]): Int = {
    def toggleLights(buttons_to_push: Array[Array[Int]]): Array[Boolean] = {
        val lights = Array.fill(lights_target.length)(false)
        for (button <- buttons_to_push; b <- button) lights(b) = !lights(b)
        lights
    }
    Range(1, buttons.length + 1)
        .find(buttons
            .combinations(_)
            .exists(toggleLights(_) sameElements lights_target))
        .get
}
def solve_z3(machine: Machine): Long = {
    val ctx = new Context()
    val solver = ctx.mkOptimize()
    val numButtons = machine.buttons.length
    val numJoltages = machine.joltages.length
    // x(i) = number of times button i is pressed
    val x = Array.tabulate(numButtons)(i => ctx.mkIntConst(s"x$i"))
    // x(i) >= 0
    for (xi <- x) solver.Add(ctx.mkGe(xi, ctx.mkInt(0)))
    // M * x = j: for each joltage row, sum of x(col) where button col affects row == joltage(row)
    for (row <- 0 until numJoltages) {
        val terms = (0 until numButtons).collect {
            case col if machine.buttons(col).contains(row) => x(col)
        }
        val sum = if (terms.size == 1) terms.head
                  else ctx.mkAdd(terms*)
        solver.Add(ctx.mkEq(sum, ctx.mkInt(machine.joltages(row))))
    }
    // minimize sum(x)
    solver.MkMinimize(ctx.mkAdd(x*))
    solver.Check()
    val model = solver.getModel
    val result = x.map(xi => model.eval(xi, true).toString.toLong).sum
    ctx.close()
    result
}

def main(args: Array[String]): Unit = {
    val machines = scala.io.Source.fromFile("src/main/scala/input10")
        .getLines
        .map(Machine(_))
        .toList
    val part1 = machines.map(m => solve_subsets(m.lights, m.buttons)).sum
    println("Part 1: " + part1)
    val part2 = machines.map(solve_z3).sum
    println("Part 2: " + part2)
}
